importer/likutim/create_unit.go

```
@@ -0,0 +1,186 @@
+			continue
+		}
+		cukn, err := models.ContentUnits(c.mdb,
+			qm.Where("id = ?", fbase.ContentUnitID.Int64),
+			qm.Load("Files"),
+		).One()
+		if err != nil {
+			log.Errorf("cant find base unit by file: %v. Error: %s", fbase, err)
+			continue
+		}
+
+		for _, uid := range d.Doubles {
+			tx, err := c.mdb.Begin()
+			if err != nil {
+				log.Errorf("problem open transaction. Error: %s", err)
+				tx.Rollback()
+				continue
+			}
+			f, err := models.Files(c.mdb, qm.Where("uid = ?", uid)).One()
+			if err != nil {
+				log.Errorf("cant find file by uid: %s. Error: %s", uid, err)
+				tx.Rollback()
+				continue
+			}
+
+			u, err := models.ContentUnits(c.mdb, qm.Where("id = ?", f.ContentUnitID.Int64)).One()
+			if err != nil {
+				log.Errorf("cant find unit by file: %v. Error: %s", f, err)
+				tx.Rollback()
+				continue
+			}
+
+			cuo, err := FindOrigin(c.mdb, u.ID)
+			if err != nil {
+				log.Errorf("cant find origin unit by unit: %v. Error: %s", u, err)
+				tx.Rollback()
+				continue
+			}
+
+			cu, err := c.createCU(cuo)
+			if err != nil {
+				log.Errorf("cant create unit %v to unit: %v. Error: %s", cu, cuo, err)
+				tx.Rollback()
+				continue
+			}
+			c.moveFiles(cukn, &cu)
+
+			d := &models.ContentUnitDerivation{
+				SourceID:  cuo.ID,
+				DerivedID: cu.ID,
+			}
+			err = cuo.AddSourceContentUnitDerivations(c.mdb, true, d)
+			if err != nil {
+				log.Errorf("cant derive unit %v to unit: %v. Error: %s", u, cuo, err)
+				tx.Rollback()
+				continue
+			}
+			err = tx.Commit()
+			if err != nil {
+				log.Errorf("problem commit transaction. Error: %s", err)
+			}
+			log.Debugf("End create new unit type LIKUTIM id: %d", cu.ID)
+		}
+	}
+}
+
+func (c *CreateUnits) moveFiles(cukm, newu *models.ContentUnit) {
+	for _, f := range cukm.R.Files {
+		if !strings.Contains(f.Name, ".doc") {
+			continue
+		}
+		f.ContentUnitID = null.Int64{Int64: newu.ID, Valid: true}
+		err := f.Update(c.mdb, "content_unit_id")
```

This is a db interaction, it can fail. You should not hide the error from the caller.
He needs to rollback the transaction on errors.